src/terminal/terminalfunctions.cc

```diff
@@ -588,6 +588,45 @@ static void CSI_DECSTR( Framebuffer* fb, Dispatcher* dispatch __attribute( ( unu
 
 static Function func_CSI_DECSTR( CSI, "!p", CSI_DECSTR );
 
+static bool Parse_OSC_8( const std::vector<wchar_t>& osc8_vector, std::string& osc8_str )
+{
+  osc8_str.reserve( osc8_vector.size() );
+  for ( wchar_t wide_char : osc8_vector ) {
+    // Valid char range is 32-126
+    if ( wide_char < 32 || wide_char > 126 ) {
+      return false;
+    }
+    osc8_str.append( 1, static_cast<char>( wide_char ) );
+  }
+  return true;
+}
+
+static void OSC_8( const std::string& OSC_string, Framebuffer* fb )
+{
+  // OSC of the form "\033]8;params;url\007"
+  assert( OSC_string[0] == '8' );
+  if ( OSC_string.size() <= 2 || OSC_string[1] != ';' ) {
+    // Bail early if the string is malformed.
+    return;
+  }
+
+  size_t second_semicolon = OSC_string.find_first_of( ';', 2 );
+  if ( second_semicolon == std::string::npos ) {
+    // Missing the second semicolon, malformed.
+    return;
+  }
+
+  std::string id;
+  std::string params = OSC_string.substr( 2, second_semicolon - 2 );
+  size_t id_pos = params.find( "id=" );
+  if ( id_pos != std::string::npos ) {
+    id = params.substr( id_pos + 3, params.find( ":", id_pos + 3 ) );
+  }
```

Comment:
I'm not sure this right.  Consider what happens if there's two key/value pairs in `params`:

`id=123:foo=bar`

`id_pos` is 0, and `params.find(":", id_pos + 3)` is 6.

So then id becomes `params.substr(3, 6)` which is `123:fo`

(The second argument to `std::string::substr` is length of the substring, not the final character position)